### PR TITLE
Extract payment function

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Extraction of payment code in CRM_Contribute_Form_Confirm to make it easier to understand/refactor.

Before
----------------------------------------
`processConfirm()` is a big long function that is difficult to follow.

After
----------------------------------------
It's a bit easier to understand and a chunk of code is extracted to make it easier to see inputs/outputs.

Technical Details
----------------------------------------

Comments
----------------------------------------
I'm trying to debug a payment failure that leaves no trace on the server and got very lost in this function.

### CODE REVIEW

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 1]
```diff
@@ -2444,8 +2444,7 @@ protected function completeTransaction($result, $contributionID) {
    */
   protected function bounceOnError($message): void {
     CRM_Core_Session::singleton()
-      ->setStatus(ts('Payment Processor Error message :') .
-        $message);
+      ->setStatus(ts('Payment Processor Error message') . ': ' . $message);
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php:2447]: This just simplifies translation

File: CRM/Core/Error.php [hunk 2]
```diff
@@ -123,7 +123,7 @@ public static function getMessages(&$error, $separator = '<br />') {
   public static function displaySessionError(&$error, $separator = '<br />') {
     $message = self::getMessages($error, $separator);
     if ($message) {
-      $status = ts("Payment Processor Error message") . "{$separator} $message";
+      $status = ts('Payment Processor Error message') . "{$separator} $message";
```
Comment (mattwire) [inline CRM/Core/Error.php:126]: Minor cleanup

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 3]
```diff
@@ -2655,6 +2618,55 @@ public function processConfirm(
...
+   * @internal
+   *
+   * @param \CRM_Contribute_DAO_Contribution $contribution
+   * @param int $contactID
+   *
+   * @return array
+   * @throws \CRM_Core_Exception
+   */
+  private function processConfirmPayment(\CRM_Contribute_DAO_Contribution $contribution, int $contactID): array {
+    $form = $this;
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php:2633]: Should do a find/replace in this function but want to do a straight extraction first.

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 4]
```diff
@@ -2655,6 +2618,55 @@ public function processConfirm(
...
+      if (!empty($paymentParams['contributionID'])) {
+        CRM_Contribute_BAO_Contribution::failPayment($paymentParams['contributionID'],
+          $paymentParams['contactID'], $e->getMessage());
+      }
+      if (!empty($paymentParams['contributionRecurID'])) {
+        CRM_Contribute_BAO_ContributionRecur::deleteRecurContribution($paymentParams['contributionRecurID']);
+      }
+
+      $result['is_payment_failure'] = TRUE;
+      $result['error'] = $e;
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php:2665]: I don't like this at all and would prefer that an exception was thrown

**Comment by eileenmcnaughton** [comment]
@mattwire this looks like a straight extraction but there is something going on in here I can't spot



api_v3_ContributionPageTest::testSubmitRecurMultiProcessorInstantPayment
Failure in api call for Contribution_Page submit:  Property 'amount' has not been set.
#0 /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/Civi/Payment/PropertyBag.php(465): Civi\Payment\PropertyBag->get('amount', 'default')
#1 /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/CRM/Core/Payment/Dummy.php(208): Civi\Payment\PropertyBag->getAmount()
#2 /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/CRM/Contribute/Form/Contribution/Confirm.php(2642): CRM_Core_Payment_Dummy->doPayment(Array)
#3 /home/homer/buildkit/build/build-1/web/sites/all/modules/civicrm/CRM/Contribute/Form/Contribution/Confirm.php(2600): CRM_Contribute_Form_Contribution_Confirm->processConfirmPayment(Object(CRM_Contribute_BAO_Contribution), 4)

**Comment by eileenmcnaughton** [comment]
Also @mattwire FYI - out of scope here but about where things might be going - I generally try to pass in an array of the parameters a function uses when I can - rather than using `$form->_params` - because it makes it more visible / explicit what values are actually used & reduces the risk of random stuff being stuff in one array to get them out at some unpredictable later point

**Comment by mattwire** [comment]
@eileenmcnaughton It was the `$paymentParams` array - that needed passing in too.
Agree about passing in array of parameters where possible, not sure when I'll get round to further refactoring but the nice thing about extracting this bit is that it removes a big chunk of code from the main function and it does an early return so we don't need to worry about anything that goes on in this bit of code affecting later down the function.

**Pull request merged**